models/eprop_iaf_psc_delta.cpp

```diff
@@ -0,0 +1,518 @@
+
+      S_.z_ = 1.0;
+    }
+
+    emplace_new_eprop_history_entry( t );
+    write_surrogate_gradient_to_history( t, S_.surrogate_gradient_ );
+    write_firing_rate_reg_to_history( t, S_.z_, P_.f_target_, P_.kappa_, P_.c_reg_ );
+
+    S_.learning_signal_ = get_learning_signal_from_history( t, false );
+
+    // set new input current
+    S_.y0_ = B_.currents_.get_value( lag );
+
+    // voltage logging
+    B_.logger_.record_data( origin.get_steps() + lag );
+  }
+}
+
+void
+nest::eprop_iaf_psc_delta::handle( SpikeEvent& e )
+{
+  assert( e.get_delay_steps() > 0 );
+
+  // EX: We must compute the arrival time of the incoming spike
+  //     explicity, since it depends on delay and offset within
+  //     the update cycle.  The way it is done here works, but
+  //     is clumsy and should be improved.
+  B_.spikes_.add_value(
+    e.get_rel_delivery_steps( kernel().simulation_manager.get_slice_origin() ), e.get_weight() * e.get_multiplicity() );
+}
+
+void
+nest::eprop_iaf_psc_delta::handle( CurrentEvent& e )
+{
+  assert( e.get_delay_steps() > 0 );
+
+  const double c = e.get_current();
+  const double w = e.get_weight();
+
+  // add weighted current; HEP 2002-10-04
+  B_.currents_.add_value( e.get_rel_delivery_steps( kernel().simulation_manager.get_slice_origin() ), w * c );
+}
+
+void
+eprop_iaf_psc_delta::handle( LearningSignalConnectionEvent& e )
+{
+  for ( auto it_event = e.begin(); it_event != e.end(); )
+  {
+    const long time_step = e.get_stamp().get_steps();
+    const double weight = e.get_weight();
+    const double error_signal = e.get_coeffvalue( it_event ); // get_coeffvalue advances iterator
+    const double learning_signal = weight * error_signal;
+
+    write_learning_signal_to_history( time_step, learning_signal, false );
+  }
+}
+
+void
+nest::eprop_iaf_psc_delta::handle( DataLoggingRequest& e )
+{
+  B_.logger_.handle( e );
+}
+
+void
+eprop_iaf_psc_delta::compute_gradient( const long t_spike,
+  const long t_spike_previous,
+  double& z_previous_buffer,
+  double& z_bar,
+  double& e_bar,
+  double& epsilon,
+  double& weight,
+  const CommonSynapseProperties& cp,
+  WeightOptimizer* optimizer )
+{
+  double e = 0.0;                // eligibility trace
+  double z = 0.0;                // spiking variable
+  double z_current_buffer = 1.0; // buffer containing the spike that triggered the current integration
+  double psi = 0.0;              // surrogate gradient
+  double L = 0.0;                // learning signal
+  double grad = 0.0;             // gradient
+
+  const EpropSynapseCommonProperties& ecp = static_cast< const EpropSynapseCommonProperties& >( cp );
+  const auto optimize_each_step = ( *ecp.optimizer_cp_ ).optimize_each_step_;
+
+  auto eprop_hist_it = get_eprop_history( t_spike_previous - 1 );
+
+  const long t_compute_until = std::min( t_spike_previous + P_.eprop_isi_trace_cutoff_, t_spike );
+
+  for ( long t = t_spike_previous; t < t_compute_until; ++t, ++eprop_hist_it )
+  {
```

Comment:
We don't need the value of `t` here and adding the cutoff to `t_spike_previous` causes slight problems, see above. Therefore, I would suggest
```suggestion
  for ( long num_steps = std::min( t_spike - t_spike_previous, V_.eprop_isi_trace_cutoff_steps_ ) ;
        num_steps > 0 ; --num_steps, ++eprop_hist_it )
  {
      assert( eprop_hist_it != eprop_history_.end() );
```
I added the assertion for to be absolutely certain that we do not run over the end of the history.

This applies correspondingly to other models.

Reply:
`get_eprop_history` is implemented with `std::lower_bound` such that it never goes beyond `eprop_history_.end()`.

Reply:
Please see https://github.com/JesusEV/nest-simulator/pull/36